src/v/storage/index_state.cc

```diff
@@ -193,6 +195,10 @@ void read_nested(
     read_nested(p, st.relative_offset_index, 0U);
     read_nested(p, st.relative_time_index, 0U);
     read_nested(p, st.position_index, 0U);
+
+    if (compat_version < index_state::offset_timestamps_serde_version) {
+        st.apply_offset_to_relative_time_index();
```

Comment:
> When we serialize this, we are always serializing with the latest version, right?

That's right.

>  That would be fine with normal serde rules about appending fields (old redpandas would just ignore trailing data), but it's a problem here

Very good point. I'm guessing you are referring to a scenario where we'd like to go back to the previous version during an upgrade (we don't support explicit downgrades). That would be problematic. I was trying to avoid a feature flag here, but I think we need to wait for the entire cluster to be upgraded before persisting to disk with the new semantics.

Reply:
It was a pain, but it had to be done: see https://github.com/redpanda-data/redpanda/commit/e59883899287867499f2eb725955eea393d69bd0 and https://github.com/redpanda-data/redpanda/commit/e7eea26390f86213746a8fe728c73bc83bf59d62